src/items/galleon.rs

```diff
@@ -0,0 +1,317 @@
+use macroquad::{
+    color,
+    experimental::{
+        animation::{AnimatedSprite, Animation},
+        collections::storage,
+        coroutines::{start_coroutine, Coroutine},
+        scene::{Handle, HandleUntyped, Node, RefMut},
+    },
+    prelude::*,
+    rand::gen_range,
+};
+
+use crate::{
+    capabilities,
+    components::{GunlikeAnimation, PhysicsBody, ThrowableItem},
+    nodes::Player,
+    Resources,
+};
+
+pub struct FlyingGalleon {
+    pub sprite: Texture2D,
+    pub pos: Vec2,
+    pub speed: Vec2,
+    pub lived: f32,
+    pub facing: bool,
+    pub owner_id: u8,
+}
+
+impl FlyingGalleon {
+    pub const SPEED: f32 = 200.;
+    pub const LIFETIME: f32 = 15.;
+    pub const WIDTH: f32 = 425.;
+    pub const HEIGHT: f32 = 390.;
+
+    pub fn new(owner_id: u8) -> FlyingGalleon {
+        let resources = storage::get::<Resources>();
+        let sprite = resources.items_textures["galleon/flying_galleon"];
+
+        let (pos, facing) = Self::start_position();
+        let dir = if facing {
+            Vec2::new(1., 0.)
+        } else {
+            Vec2::new(-1., 0.)
+        };
+
+        FlyingGalleon {
+            sprite,
+            pos,
+            speed: dir * Self::SPEED,
+            lived: 0.,
+            facing,
+            owner_id,
+        }
+    }
+
+    pub fn start_position() -> (Vec2, bool) {
+        let resources = storage::get::<Resources>();
+        let map_width =
+            resources.tiled_map.raw_tiled_map.tilewidth * resources.tiled_map.raw_tiled_map.width;
+        let map_height =
+            resources.tiled_map.raw_tiled_map.tileheight * resources.tiled_map.raw_tiled_map.height;
+
+        let facing = gen_range(0, 2) == 0;
+        let pos = Vec2::new(
+            if facing {
+                -Self::WIDTH
+            } else {
+                map_width as f32
+            },
+            gen_range(0., map_height as f32 - Self::HEIGHT),
+        );
+
+        (pos, facing)
+    }
+
+    pub fn update(&mut self) -> bool {
+        self.pos += self.speed * get_frame_time();
+        self.lived += get_frame_time();
+
+        if self.lived > Self::LIFETIME {
```

Comment:
The lifetime logic can be entirely removed, by performing the check based on the position.

Pseudocode:

```
if object.x >= SCREEN_WIDTH || object.x <= - OBJECT_WIDTH {
  # object is dead
}
```

Reply:
I removed it, but I used the map width as base to removing the flying galleon.